### PR TITLE
Capability field analyzer

### PR DESCRIPTION
**Pull request by kozlovb**
Add's a public capability field analyzer

Closes #https://github.com/onflow/cadence-tools/issues/6

### CODE REVIEW

File: lint/capability_fields_analyser.go [hunk 1]
```diff
@@ -20,35 +20,69 @@ package lint
 
 import (
 	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
 	"github.com/onflow/cadence/tools/analysis"
 )
 
-func DetectCapabilityType(typeToCheck ast.Type) bool {
+func DetectCapabilityType(typeToCheck ast.Type, structWithPubCapabilities map[string]struct{}) bool {
 	const capabilityTypeName = "Capability"
```
Comment (nozim) [inline lint/capability_fields_analyser.go:28]: prob worth making it global const with Capital name, wyt? )

File: lint/capability_fields_analyser.go [hunk 2]
```diff
@@ -20,35 +20,69 @@ package lint
...
+					}
+					for _, d := range declaration.Members.Declarations() {
+						field, ok := d.(*ast.FieldDeclaration)
+						if !ok {
+							return
+						}
+						if field.Access != ast.AccessPublic {
+							return
+						}
+						if field.Access == ast.AccessPublic && DetectCapabilityType(field.TypeAnnotation.Type, structWithPubCapabilities) {
```
Comment (nozim) [inline lint/capability_fields_analyser.go:74]: if field.Access == ast.AccessPublic will allways be false because you check it before, isn't it?

File: lint/capability_fields_analyser.go [hunk 3]
```diff
@@ -74,6 +107,32 @@ var CapabilityFieldAnalyzer = (func() *analysis.Analyzer {
 					if !ok {
 						return
 					}
+
+					nt, ok := field.TypeAnnotation.Type.(*ast.NominalType)
+					if ok {
+						hasCapability, found := compositeIdsCapabilitiesPropery[nt.Identifier.Identifier]
+
+						if found {
```
Comment (nozim) [inline lint/capability_fields_analyser.go]: how about 
`if !ok { return }`
`if !found {return} 
//. the rest `

As far as I remember idiomatic go tends to avoid nested conditionals to enhance readability :)

File: lint/capability_fields_analyser.go [hunk 4]
```diff
@@ -74,6 +107,32 @@ var CapabilityFieldAnalyzer = (func() *analysis.Analyzer {
...
+							if hasCapability && field.Access == ast.AccessPublic {
```
Comment (nozim) [inline lint/capability_fields_analyser.go]: same here:
`if !hasCapability { return }`
`if field.Access != ast.AccessPublic { return }
// report here 
` ?

File: lint/capability_fields_analyser.go (hunk 1 above)
Comment (kozlovb) [reply lint/capability_fields_analyser.go:28]: I don't think that it needs to be global.

File: lint/capability_fields_analyser.go (hunk 3 above)
Comment (kozlovb) [reply lint/capability_fields_analyser.go]: I agree with your comment, but I don't have this code on my machine. And I've just checked I don't see uncommitted changes. Let me push again my branch to GitHub and see.

Comment (kozlovb) [reply lint/capability_fields_analyser.go]: I agree with your comments but somehow I don't see this code on my machine. And when Ii try to push my branch nothing changes it says that it's up to date ... 


I'll try push force now.

File: lint/capability_fields_analyser.go (hunk 2 above)
Comment (kozlovb) [reply lint/capability_fields_analyser.go:74]: thx, you are correct its superfluous, I had it on my local machine's branch too.

File: lint/capability_fields_analyser.go (hunk 3 above)
Comment (kozlovb) [reply lint/capability_fields_analyser.go]: Yeah, seems like I forgot to push or needed to do a push force. Please accept my apologies, you would need to have another look as code is a bit different.

File: lint/capability_fields_analyser.go (hunk 4 above)
Comment (kozlovb) [reply lint/capability_fields_analyser.go]: Please refer to this comment https://github.com/onflow/cadence-tools/pull/176#discussion_r1272119853

**Comment by turbolent** [comment]
@kozlovb Just to clarify: Does that mean that this PR subsumes #163 and #164, i.e., should those two other PRs be closed?

**Comment by kozlovb** [comment]
@turbolent thx for getting back to us. Yes, I believe so. @nozim is a co-author of this one. @nozim pls confirm that those can be closed as this one has your work merged with mine.

**Comment by nozim** [comment]
Yes I believe so, @kozlovb already implemented those cases in his branch.

File: lint/capability_fields_analyser.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+					}
+					for _, d := range declaration.Members.Declarations() {
+						field, ok := d.(*ast.FieldDeclaration)
+						if !ok {
+							return
+						}
+						if field.Access != ast.AccessPublic {
+							return
+						}
+						if DetectCapabilityType(field.TypeAnnotation.Type, structWithPubCapabilities) {
```
Comment (turbolent) [inline lint/capability_fields_analyser.go:74]: Currently, `DetectCapabilityType` detects a type which is or includes the `Capability` type purely on the syntax tree (`ast.Type`). Maybe consider operating on the semantic level, i.e. `sema.Type`.

File: lint/capability_fields_analyser_test.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+		require.Equal(
+			t,
+			[]analysis.Diagnostic(nil),
+			diagnostics,
+		)
+	})
+}
+
+func TestImplicitCapabilityLeakViaStruct(t *testing.T) {
+	t.Run("leak via struct field", func(t *testing.T) {
```
Comment (turbolent) [inline lint/capability_fields_analyser_test.go:228]:
```suggestion
	t.Parallel()

	t.Run("leak via struct field", func(t *testing.T) {
```

File: lint/capability_fields_analyser_test.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+					Location:         testLocation,
+					Category:         lint.UpdateCategory,
+					Message:          "It is an anti-pattern to have public Capability fields.",
+					SecondaryMessage: "Consider restricting access.",
+				},
+			},
+			diagnostics,
+		)
+	})
+	t.Run("valid", func(t *testing.T) {
```
Comment (turbolent) [inline lint/capability_fields_analyser_test.go:285]:
```suggestion
	})

	t.Run("valid", func(t *testing.T) {
```

File: lint/capability_fields_analyser.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+func CollectStructsWithPublicCapabilities(inspector *ast.Inspector) (map[string]struct{}, map[ast.Identifier]struct{}) {
+	structWithPubCapabilities := make(map[string]struct{})
+	fieldsInStruct := make(map[ast.Identifier]struct{})
+	inspector.Preorder(
+		[]ast.Element{(*ast.CompositeDeclaration)(nil)},
+		func(element ast.Element) {
+			switch declaration := element.(type) {
+			case *ast.CompositeDeclaration:
+				{
+					if declaration.CompositeKind != common.CompositeKindStructure {
```
[the remark this answers is not in the document]
Comment (kozlovb) [reply lint/capability_fields_analyser.go:63]: Good question. My thought process was the following ( I mean I saw the idea of looking into structs in the Nozim's patch) -
Structures can be used as types to declare fields in contracts. So, in that way we would have a map of the structs that have a potentially dangerous public capabilities that might leak. I am referring to this test -
```
			`
 		pub contract MyContract {
 		   pub resource Counter {
 		     priv var count: Int
 		
 		     init(count: Int) {
 		       self.count = count
 		     }
 		   }
 		
 		   pub struct ContractData {
 			   pub var owner: Capability
 		       init(cap: Capability) {
 		           self.owner = cap
 		       }
 		   }

			pub struct ContractDataIgnored {
 			   pub var owner: Capability
 		       init(cap: Capability) {
 		           self.owner = cap
 		       }
 		   }
 		
 		   pub var contractData: ContractData
 		
 		   init(){
 		       self.contractData = ContractData(cap:
 		           self.account.getCapability<&Counter>(/public/counter)
 		       )
 		   }
 		}`,
```

I think for contracts it doesn't make sense to have such a map. Or is it? I have my day off on Friday so can take a more in depth look.

For resources you are probably right as a resource might contain a public capability which links to the other resource. I need to check Cadence a bit as it's been a while since I looked into it.

**Comment by kozlovb** [comment]
@turbolent thx for the review. I'll take a more in depth look on Friday. Otherwise, I had 2 question and for everything else I'll just apply your suggestions.

File: lint/capability_fields_analyser.go (hunk 5 above)
Comment (turbolent) [reply lint/capability_fields_analyser.go:74]: The same can be done for `sema.Type`, e.g. there's `sema.ConstantSizedType`, `sema.DictionaryType`, etc.

Comment (kozlovb) [reply lint/capability_fields_analyser.go:74]: I tried to take a look today. I see that usually ast is used when checks are simple like in reference_operator_analyser.go etc... I found in redundant_cast_analyser.go sema is used. 
The conversion from ast to sema is done like 
```redundantType := elaboration.StaticCastTypes(castingExpression)```
where ```castingExpression``` is of ```*ast.CastingExpression``` type. 

So, I went to elaboration.go but  couldn't find any function like StaticCastTypes that would take ```*ast.FieldDeclaration``` as an argument.  And I'd need to elaborate on a declaration.
Do you think here I an easy way to get from ```*ast.FieldDeclaration``` to relevant ```sema.Type```s ?

Comment (kozlovb) [reply lint/capability_fields_analyser.go:74]: Hi @turbolent , so what do you think about my previous comment ? I couldn't find an existing function in elaboration that would allow me to convert so sema.

Comment (turbolent) [reply lint/capability_fields_analyser.go:74]: Sorry for dropping the ball here Boris.

It's OK to leave it as-is for now. Given there's no shadowing allowed fore types, it should be sufficient to detect `Capability` purely syntactically (i.e. through the AST instead of sema).

File: lint/capability_fields_analyser.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (turbolent) [reply lint/capability_fields_analyser.go:63]: That makes sense. Given that a public capability fields is just as problematic in a resource or contract as it is in a struct, it might make sense to just detect it everywhere, i.e. remove the if-statement.

File: lint/capability_fields_analyser.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+package lint
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/tools/analysis"
+)
+
+func DetectCapabilityType(typeToCheck ast.Type, structWithPubCapabilities map[string]struct{}) bool {
```
[the remark this answers is not in the document]
Comment (kozlovb) [reply lint/capability_fields_analyser.go:27]: Agreed, will do.

Not cause I am nitpicking, but just to clarify -

The idea is that contract is checked as a whole and once it has a "leak" warning is issued.
However, for structs and resources it's different. It's fine to have a struct with a public capability field if this struct is always used to declare private fields in a contract.

For this reason a map of structWithPubCapabilities is constructed. It will allow to see if those structs are used to declare public fields. For contracts - we won't include them in this map as once there is a contract with a leak we already issue a warning. 

TBH I dont know if contract can be embedded in another contract. Seems, like this is not the case in Cadence. So, also a reason why contract shouldn't be in this map. Pls correct my understanding , my kwnoledge of Cadence is on a beginner level.

File: lint/capability_fields_analyser.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+	inspector.Preorder(
+		[]ast.Element{(*ast.CompositeDeclaration)(nil)},
+		func(element ast.Element) {
+			switch declaration := element.(type) {
+			case *ast.CompositeDeclaration:
+				{
+					if declaration.CompositeKind != common.CompositeKindStructure {
+						return
+					}
+					for _, d := range declaration.Members.Declarations() {
```
[the remark this answers is not in the document]
Comment (kozlovb) [reply lint/capability_fields_analyser.go]: Closing thread without applying as per other comment the ```if declaration.CompositeKind != common.CompositeKindStructure``` was removed.

File: lint/capability_fields_analyser.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/tools/analysis"
+)
+
+func DetectCapabilityType(typeToCheck ast.Type, structWithPubCapabilities map[string]struct{}) bool {
+	const capabilityTypeName = "Capability"
+	switch downcastedType := typeToCheck.(type) {
+	case *ast.NominalType:
+		_, found := structWithPubCapabilities[downcastedType.Identifier.Identifier]
+		return downcastedType.Identifier.Identifier == capabilityTypeName || found
```
[the remark this answers is not in the document]
Comment (kozlovb) [reply lint/capability_fields_analyser.go:32]: done, thx

File: lint/capability_fields_analyser.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kozlovb) [reply lint/capability_fields_analyser.go]: I see. Then proceeding with the current implementation for now.

**Comment by kozlovb** [comment]
@turbolent I think I addressed all suggestions. Please take another look.

File: lint/capability_fields_analyser.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+	case *ast.ConstantSizedType:
+		return DetectCapabilityType(downcastedType.Type, compositesWithPubCapabilities)
+	case *ast.DictionaryType:
+		return DetectCapabilityType(downcastedType.KeyType, compositesWithPubCapabilities) || DetectCapabilityType(downcastedType.ValueType, compositesWithPubCapabilities)
+	case *ast.FunctionType:
+		return false
+	case *ast.ReferenceType:
+		return DetectCapabilityType(downcastedType.Type, compositesWithPubCapabilities)
+	case *ast.RestrictedType:
+		return false
```
Comment (SupunS) [inline lint/capability_fields_analyser.go:46]: I think the inner types of the restricted type would also need to be visited recursively.
e.g:
```cadence
pub contract MyContract {
	
    pub var contractData: ContractData{SomeInterface}  //  <-- A restricted type, whose inner type has a public capability field.

    pub struct ContractData: SomeInterface {
        pub var owner: Capability

        init(cap: Capability) {
            self.owner = cap
        }
    }

    pub struct interface SomeInterface { 
        pub var owner: Capability
    }

}
```

File: lint/capability_fields_analyser.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+			structTypesPublicCapability, fieldsInStruct := CollectCompositesWithPublicCapabilities(inspector)
+
+			inspector.Preorder(
+				elementFilter,
+				func(element ast.Element) {
+					field, ok := element.(*ast.FieldDeclaration)
+					if !ok {
+						return
+					}
+					_, found := fieldsInStruct[field.Identifier]
```
Comment (SupunS) [inline lint/capability_fields_analyser.go:110]: Skipping based on the name can be a bit problematic. For e.g. what if both the contract and the struct/resource have fields with the same name? Can you maybe add a test?

Comment (kozlovb) [reply lint/capability_fields_analyser.go:110]: true, I'll take a look at it today. Probably, a more elaborated keyword be required like name+position or something of a sort.

Comment (SupunS) [reply lint/capability_fields_analyser.go:110]: Or maybe instead of doing it in two iterations, perform this check in a single iteration.
i.e:
- Visit each field of the contract
- Check if the type of the field has public capability fields:
  - For that, visit the type then and there, recursively.
  - Once the result is found memoize it so it can be looked up if the same type is encountered again. (in the memoized mapping, Type could be the key, and the result/boolean could be the value)
- It might be easier to use the sema-type instead.

I haven't actually tried this at the code level, but just an idea that might be easier to implement.

Comment (kozlovb) [reply lint/capability_fields_analyser.go:110]: Sorry, finally didn't have time on Monday, now I am aiming Friday.  (I work full time Tue, Wed, Thu)

As for using sema we had this discus with Bastian, pls take a look at my comment 
https://github.com/onflow/cadence-tools/pull/176#discussion_r1284688013

As for your proposal regarding one path. 
"Once the result is found memoize it so it can be looked up if the same type is encountered again. (in the memoized mapping, Type could be the key, and the result/boolean could be the value)" - I think it's a valid optimization that would allow to limit  number of runs of ```DetectCapabilityType```.  Although, may be there ll be difficulties in case of mane collisions that you ve mentioned in your previous comment. ( the rick with key - name+location wont work here I think)

However, I needed 2 pathes for a different reason.

1st path detects composites , so resources and struct with public capability type. However , it's not a bad pattern yet to have them declared like this. if this struct is only used as a private field in a contract that's fine.
2nd path well ignores pub capabilities inside of structs and resources but checks them in contracts and also takes structs/resources with pub capabilities from 1st path and checks if those are used to declare public fields.

Basically, pls take a look at this test - 
```
			`
 			pub contract MyContract {
 			   pub resource Counter {
 				 priv var count: Int
 			
 				 init(count: Int) {
 				   self.count = count
 				 }
 			   }
 		
 			   pub struct ContractData {
 				   pub var owner: Capability
 				   init(cap: Capability) {
 					   self.owner = cap
 				   }
 			   }
 		
                priv var contractData: ContractData
 		
 			   init(){
 				   self.contractData = ContractData(cap:
 					   self.account.getCapability<&Counter>(/public/counter)
 				   )
 			   }
 			}
 			`
```
it was reporting a false warning before I introduced 2 paths. I just couldn't come up with a solution that uses one path.

Comment (SupunS) [reply lint/capability_fields_analyser.go:110]: >may be there ll be difficulties in case of name collisions

In the algorithm I suggested, there will be no name collisions since field names are not looked up, they are visiting in-place.

> it was reporting a false warning before I introduced 2 paths. I just couldn't come up with a solution that uses one path.
Here, it approach I suggested would not report any errors. Since it first (and only) visit the field `priv var contractData: ContractData`, sees that field is private, so skips.
Suppose the field was public, then only it will go and check the `ContractData ` type.

Comment (kozlovb) [reply lint/capability_fields_analyser.go:110]: but how it would skip 
```
    pub struct ContractData {
        pub var owner: Capability
            init(cap: Capability) {
                 self.owner = cap
            }
     }
```
?

Comment (SupunS) [reply lint/capability_fields_analyser.go:110]: Like I mentioned in https://github.com/onflow/cadence-tools/pull/176#discussion_r1316381572, we would only need to visit contract fields, no need to visit type definitions.

Comment (kozlovb) [reply lint/capability_fields_analyser.go:110]: Also, how would it visit ```ContractData``` if needed ? 
May be there is a way to run ```inspector.Preorder( ...``` on a specific struct but I didn't find an example.
Well, I am not saying that one path wont work but I just fail to see how to achieve it with my current knowledge.

Comment (kozlovb) [reply lint/capability_fields_analyser.go:110]: "Like I mentioned in https://github.com/onflow/cadence-tools/pull/176#discussion_r1316381572, we would only need to visit contract fields, no need to visit type definitions, unless it is encountered as a type of a public field."

ah sorry I didn't understand that you proposed to ignore definitions.
I don't know how to ignore type definitions but I imagine there should be a way to filter them out together with resources. However, if I would need to analyze ```ContractData``` how I can make a path through it ?
Is there a way to run an inspector on a specific struct definition ? 

Other thing, the advantage of having two paths - all structs with pub capabilities are stored in a map so there ll be no need to analyze them on each encounter.

Comment (SupunS) [reply lint/capability_fields_analyser.go:110]: `pass.Program` already has the `ast` nodes as well as type info to look up. So when you know a type (say, `ContractData`), it can be retrieved from the `pass.Program`.

File: lint/capability_fields_analyser.go (hunk 5 above)
Comment (SupunS) [reply lint/capability_fields_analyser.go]: We might have to do this using the sema-type, because the contract could be using types that are imported, and for them, the ast-node (`ast.Type`) might not be available at this stage.
e.g:
```cadence
import ImprtedContract from 0x01

pub contract MyContract {

    // Assume `ImprtedContract.ContractData` has a public capability field.
    // It won't be able to report an error here only by looking at the syntax-tree
    pub var contractData: ImprtedContract.ContractData
```

File: lint/capability_fields_analyser.go (hunk 13 above)
Comment (kozlovb) [reply lint/capability_fields_analyser.go:110]: thx for the suggestion. For now I don't understand how to transit from 
ast-type-definition (in ```pass.Program.Program```) to the "reduced" inspector so to be able to run 

 ```func CollectCompositesWithPublicCapabilities(inspector *ast.Inspector) (map[string]struct{}, map[ast.Identifier]struct{}) {```

on an inspector that only is a result of this specific node. But I'll take a look.

Other note - anyway I think caching in the map is nice as this would allow not to analyze the same structs if used to declare several fields in different places of a contract.

**Comment by kozlovb** [comment]
Hi @SupunS , I worked today on a solution with one pass. Indeed, I could use `pass.Program.Program.CompositeDeclarations()` to get declarations, may be later I could use `sema` as well. 


However, I dont understand how to ignore definitions while performing 
`func (in *Inspector) Preorder(types []Element, f func(Element))`

What did you have in mind when you commented - 

"Like I mentioned in https://github.com/onflow/cadence-tools/pull/176#discussion_r1316381572, we would only need to visit contract fields, no need to visit type definitions, unless it is encountered as a type of a public field."

So, how to avoid to visit fields declaration inside of type/struct definitions ? 
Shell I try to use something else than Preorder ?

**Comment by turbolent** [comment]
Replaced by #551
